### PR TITLE
add end-to-end example gallery for transforms v2

### PR DESCRIPTION
**Pull request by pmeier**
cc @vfdev-5 @bjuncek

### CODE REVIEW

File: gallery/assets/coco/images/000000000001.jpg
Comment (NicolasHug) [inline gallery/assets/coco/images/000000000001.jpg:1]: Is that a `.jpg` file that contains text? Or is GitHub tricking me?
Is the absolute path (`/home/philip/git/...`) going to break?

Comment (pmeier) [reply gallery/assets/coco/images/000000000001.jpg:1]: No it was a symlink to avoid duplicating the image. But I guess that didn't work out. Let me see how it should be done for git.

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:18]:
```suggestion
# We are using BETA APIs, so we deactivate the associated warning,
# thereby acknowledging that some APIs may slightly change in the future
torchvision.disable_beta_transforms_warning()
```

File: gallery/assets/coco/images/000000000001.jpg
Comment (pmeier) [reply gallery/assets/coco/images/000000000001.jpg]: Fixed with relative symlink

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:45]: Can we put the `torchvision.disable_beta_transforms_warning()` and the associated comment just above this line?

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:42]: ~have~?

or "have to" ?

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:53]:
```suggestion
    # We're using a toy dataset with the same coco structure,
    # in practice you'll have to replace this with the proper dataset root.
    root = pathlib.Path("assets") / "coco"
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:73]: It's not clear from this output that the type of the dict entries have been changed, since we're just showing the keys

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:79]:
```suggestion
# Let's define an object detection pipeline
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:97]: Lol... should we fix that before we merge to not risk that kind of comment in the published docs?

**Review by NicolasHug — COMMENTED** [review]
First pass, looks good so far

**Comment by pmeier** [comment]
[Rendered](https://output.circle-artifacts.com/output/job/795c628e-5fba-46d3-a907-ea370aa26a3c/artifacts/0/docs/auto_examples/plot_transforms_v2_e2e.html#sphx-glr-auto-examples-plot-transforms-v2-e2e-py)

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [reply gallery/plot_transforms_v2_e2e.py:111]: Opened https://github.com/pytorch/vision/pull/7304

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:43]: Looking at this more we could consider showing the warning (which would have links to the relevant feedback issues) and just mention that one can call `disable_beta_transforms_warning()` to disable it, but no strong opinion.

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:3]: I don't know how to keep it short, but it'd be nice to emphasize in the title that one can now jointly transforms images/boxes/masks.

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:39]: Not sure if this was missed before but this sentences looks slightly weird. Below is one of many alternative suggestions

```suggestion
# To use ``transforms.v2``, you can use the stable datasets and models and you just need to import the transforms differently
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:74]:
```suggestion
# :func:`~torchvision.datapoints.wrap_dataset_for_transforms_v2` function. For :class:`~torchvision.datasets.CocoDetection` this changes the
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:73]:
```suggestion
# dictionaries, which each containing the annotations for a single object instance. As is, this format is not compatible
# with the transforms v2, nor with the models. To overcome that, we provide the
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:50]:
```suggestion
# We start off by loading the :class:`~torchvision.datasets.CocoDetection` dataset to have a look at what
# it currently returns, and we'll see how to convert it to a format that is compatible with our new transforms.```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:85]:
```suggestion
print(type(target["boxes"]), type(target["masks"]))
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:76]: super NIT but label isn't a datapoint anymore. Not sure it's worth changing the sentence though.

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:88]: Nit, just to avoid "look" twice

```suggestion
# As baseline, let's have a look at a sample without transformations
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:95]:
```suggestion
# transforms v1, but now handles bounding boxes and masks without any extra configuration.
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:98]:
```suggestion
#    Although the :class:`~torchvision.transforms.v2.SanitizeBoundingBoxes` transform is a no-op in this example, it should be placed at least once
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:100]:
```suggestion
#    optionally masks. It is particularly critical to add it if class:`~torchvision.transforms.v2.RandomIoUCrop` was used.
```

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:117]: > sphinx_gallery_thumbnail_number = 2

Looks like this is being rendered as pure text, consider adding it to the top as in https://github.com/pytorch/vision/blob/main/gallery/plot_optical_flow.py#L25.

File: gallery/plot_transforms_v2_e2e.py
Comment (NicolasHug) [inline gallery/plot_transforms_v2_e2e.py:129]: Maybe add something along the lines of 

> We can see that our image got reshaped, cropped and translated, and the bounding box was also transformed accordingly.

**Review by NicolasHug — APPROVED** [review]
This looks great, thanks Philip!
I have a bunch of minor comments below but nothing critical. I'll approve to unblock any follow-up, feel free to merge whenever

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py:43]: I can push a commit that shows this to have a better base for discussion. Will update here later.

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py:3]: Maybe not in the title, but in the description just below the title? Something along the lines: 

> Object detection is not supported out of the box by transforms v1, since it only supports images. transforms v2 enables jointly transforming amongst others images, bounding boxes, and masks. This example showcases an end-to-end object detection training using stable torchvision datasets and models as well as the new tranforms v2 API.

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py:74]: You mean 

```suggestion
# :func:`~torchvision.datasets.wrap_dataset_for_transforms_v2` function. For :class:`~torchvision.datasets.CocoDetection` this changes the
```

right? Recall that the wrapper is now located in the dataset namespace.

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py:76]: I thought about it too, but didn't find anything. Unless you can come up with something better, I'll leave it as is.

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py]: I went for my suggestion, although for now the link is broken anyway, since the wrapper function is not yet documented.

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py:3]: I went for my suggestion in the latest commit. LMK if you are ok with that.

**Comment by pmeier** [comment]
[Rendered](https://output.circle-artifacts.com/output/job/0cdbff79-79ad-4cfb-ad68-03bf3d844bbb/artifacts/0/docs/auto_examples/plot_transforms_v2_e2e.html#sphx-glr-auto-examples-plot-transforms-v2-e2e-py)

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py]: See 389103ffe293a880b69bca6db19383c4a5e9dd13

Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py]: Due to

https://github.com/pytorch/vision/blob/a46d97c96dfb2f7f9ddc7f4f889d9856b46428ad/docs/source/conf.py#L36

the warning is not shown even if we don't call that explicitly ourselves in the gallery again:

https://output.circle-artifacts.com/output/job/f238016d-ff53-4d71-ae1d-769bef342b86/artifacts/0/docs/auto_examples/plot_transforms_v2_e2e.html#sphx-glr-auto-examples-plot-transforms-v2-e2e-py

Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py]: Shall we merge the imports into the other ones in this case?

Comment (NicolasHug) [reply gallery/plot_transforms_v2_e2e.py]: > Shall we merge the imports into the other ones in this case?

Not sure what you mean but I'd still rather keep the call to ` torchvision.disable_beta_transforms_warning() ` and keep it close to the v2 imports. Even if it's a no-op in our doc build, it is necessary for users executing this code locally either through copy/pasting, by downloading the associated notebook, or simply executing the `.py` script directly (which is still one of the intended use-cases of the gallery)

Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py]: Ok, let's go back to disabling it explicitly and including a comment.

**Comment by pmeier** [comment]
For some reason the xref from the gallery to the documentation of our transforms is not working. They all link to https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module. However, the documentation of the transforms itself correctly recognizes that it is used in the gallery: https://output.circle-artifacts.com/output/job/0cdbff79-79ad-4cfb-ad68-03bf3d844bbb/artifacts/0/docs/generated/torchvision.transforms.v2.RandomHorizontalFlip.html#torchvision.transforms.v2.RandomHorizontalFlip Unfortunately, the v1 transform also claims to be used in the same gallery: https://output.circle-artifacts.com/output/job/0cdbff79-79ad-4cfb-ad68-03bf3d844bbb/artifacts/0/docs/generated/torchvision.transforms.RandomHorizontalFlip.html#torchvision.transforms.RandomHorizontalFlip

**Comment by NicolasHug** [comment]
That seems to be an existing problem, I observe the same for e.g. https://pytorch.org/vision/main/auto_examples/plot_transforms.html#sphx-glr-auto-examples-plot-transforms-py. Do you want to open an issue so we can keep track of this?

**Comment by pmeier** [comment]
See #7307.

File: gallery/plot_transforms_v2_e2e.py
Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py:111]: This was fixed in 85fff2c911edce21d34f3fed79418568e5fc81af and is available in the latest preview: https://output.circle-artifacts.com/output/job/f238016d-ff53-4d71-ae1d-769bef342b86/artifacts/0/docs/auto_examples/plot_transforms_v2_e2e.html#sphx-glr-auto-examples-plot-transforms-v2-e2e-py

If you are talking about the xref not working here, this is expected since we so far have no documentation for `SanitizeBoundingBoxes` or `RandomIoUCrop`.

Comment (vfdev-5) [reply gallery/plot_transforms_v2_e2e.py:111]: I was talking about `class:~torchvision...`, good as it is already fixed.

**Comment by pmeier** [comment]
[Rendered](https://output.circle-artifacts.com/output/job/3c455e7d-307b-4b66-a0cb-08d10c80278c/artifacts/0/docs/auto_examples/plot_transforms_v2_e2e.html#sphx-glr-auto-examples-plot-transforms-v2-e2e-py)

File: gallery/plot_transforms_v2_e2e.py
Comment (vfdev-5) [inline gallery/plot_transforms_v2_e2e.py:3]: Sorry for being late for the review.
```diff
- transforms v2: End-to-end object detection example
+ Transforms v2: End-to-end object detection example
```

File: gallery/plot_transforms_v2_e2e.py
Comment (vfdev-5) [inline gallery/plot_transforms_v2_e2e.py:23]: Is this necessary ? Looks a bit weird once rendered: https://output.circle-artifacts.com/output/job/a6035cf7-ba73-4f96-b92e-1b2aa5b18cfc/artifacts/0/docs/auto_examples/plot_transforms_v2_e2e.html#sphx-glr-auto-examples-plot-transforms-v2-e2e-py

Comment (NicolasHug) [reply gallery/plot_transforms_v2_e2e.py:23]: This is how we tell sphinx-gallery's to use the (last?) image for the thumbnail in this page
image

Comment (pmeier) [reply gallery/plot_transforms_v2_e2e.py:23]: I remembered that there is an option to hide these kind of comments: https://sphinx-gallery.github.io/stable/configuration.html#removing-config-comments Will send a PR

**Pull request merged**
